modules/programs/sbt.nix

```
@@ -0,0 +1,122 @@
+
+  renderCredential = cred: ''
+    credentials += Credentials("${cred.realm}", "${cred.host}", "${cred.user}", "${cred.password}")
+  '';
+
+  sbtTypes = {
+    plugin = types.submodule {
+      options = {
+        org = mkOption {
+          type = types.str;
+          description = "The organization the artifact is published under";
+        };
+        artifact = mkOption {
+          type = types.str;
+          description = "The name of the artifact";
+        };
+        version = mkOption {
+          type = types.str;
+          description = "The version of the plugin";
+        };
+      };
+    };
+
+    credential = types.submodule {
+      options = {
+        realm = mkOption {
+          type = types.str;
+          description = "The realm of the repository you're authenticating to";
+        };
+        host = mkOption {
+          type = types.str;
+          description =
+            "The hostname of the repository you're authenticating to";
+        };
+        user = mkOption {
+          type = types.str;
+          description = "The user you're using to authenticate";
+        };
+        password = mkOption {
+          type = types.str;
+          description =
+            "The password or authentication token for the repository";
+        };
+      };
+    };
+  };
+
+  cfg = config.programs.sbt;
+in {
+  meta.maintainers = [ maintainers.kubukoz ];
+
+  options.programs.sbt = {
+
+    enable = mkEnableOption "sbt";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.sbt;
+      description = "The package with sbt to be installed";
+    };
+
+    baseConfigPath = mkOption {
+      type = types.str;
+      default = ".sbt/1.0";
+      description = "Where the plugins and credentials should be located";
+    };
+
+    plugins = mkOption {
+      type = types.listOf (sbtTypes.plugin);
+      default = [ ];
+      description =
+        "A list of plugins to place in the sbt configuration directory";
+      example = [
+        {
+          org = "net.virtual-void";
+          artifact = "sbt-dependency-graph";
+          version = "0.10.0-RC1";
+        }
+        {
+          org = "com.dwijnand";
+          artifact = "sbt-project-graph";
+          version = "0.4.0";
+        }
+      ];
+      apply = v: concatStrings (map renderPlugin v);
+    };
+
+    credentials = mkOption {
+      type = types.listOf (sbtTypes.credential);
+      default = [ ];
+      description =
+        "A list of credentials to define in the sbt configuration directory";
+      example = [{
+        realm = "Sonatype Nexus Repository Manager";
+        host = "example.com";
+        user = "user";
+        password = "password";
+      }];
+      apply = v: concatStrings (map renderCredential v);
+    };
+  };
+
+  config = mkIf cfg.enable (mkMerge [
+    { home.packages = [ cfg.package ]; }
+
+    (mkIf (cfg.plugins != "") {
+      home.file."${cfg.baseConfigPath}/plugins/plugins.sbt".text = cfg.plugins;
```

Better assemble the files here, not in the `apply` function. I.e.
```suggestion
      home.file."${cfg.baseConfigPath}/plugins/plugins.sbt".text =
        concatStrings (map renderPlugin cfg.plugins);
```

---

Why? I'm genuinely curious about when I should use `apply`, then.

---

(updated)

---

The `apply` option should typically be used when you have an option that allows two or more value types and you want to normalize any set value so that when the option is read back it will always present a value of one type. In general, the apply should always produce a value that is of a valid type for the option. In the sbt plugin case, you switched types so if anybody would like to use that option they'd get an unexpected value. For example, imagine somebody wanted to create a plain text file with the names of all plugins. They'd, quite reasonably, write something like

    home.file.my-plugins.text =
        concatMapStringsSep "\n" (plugin: plugin.artifact) config.programs.sbt.plugins;

but I believe that would fail since `config.programs.sbt.plugins` was turned into a string.

A good example of what I'd consider correct use is the `passwordCommand` one I linked elsewhere.